File: bench/polyglot/cpp/sha_digest.cpp
#include "runtime/digest.h"

#include <array>
#include <cstdint>
#include <iostream>
#include <string>
#include <string_view>

namespace {

constexpr std::array<std::string_view, 4> kPayloads = {
    "Amber digest polyglot benchmark payload zero",
    "Amber digest polyglot benchmark payload one 1234567890",
    "The quick brown fox jumps over the lazy dog",
    "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
};
constexpr std::string_view kKey = "amber-digest-benchmark-key";

std::uint64_t fold_digest(std::uint64_t checksum, const std::string &digest) {
  return checksum + digest.size() + static_cast<unsigned char>(digest.front()) +
         static_cast<unsigned char>(digest.back());
}

std::uint64_t main_workload() {
  std::uint64_t checksum = 0;
  for (std::size_t i = 0; i < 4000U; ++i) {
    const std::string_view data = kPayloads[i % kPayloads.size()];
    checksum = fold_digest(checksum, amber::runtime::digest_crc32(data));
    checksum = fold_digest(checksum, amber::runtime::digest_md5(data));
    checksum = fold_digest(checksum, amber::runtime::digest_sha1(data));
    checksum = fold_digest(checksum, amber::runtime::digest_sha256(data));
    checksum =
        fold_digest(checksum, amber::runtime::digest_hmac_sha256(kKey, data));
  }
  return checksum;
}

} // namespace

int main() {
  std::cout << main_workload() << '\n';
  return 0;
}

File: bench/polyglot/go/sha_digest.go
package main

import (
	"crypto/hmac"
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"hash/crc32"
)

var payloads = [][]byte{
	[]byte("Amber digest polyglot benchmark payload zero"),
	[]byte("Amber digest polyglot benchmark payload one 1234567890"),
	[]byte("The quick brown fox jumps over the lazy dog"),
	[]byte("0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"),
}

var key = []byte("amber-digest-benchmark-key")

func foldDigest(checksum uint64, digest []byte) uint64 {
	return checksum + uint64(len(digest)) + uint64(digest[0]) +
		uint64(digest[len(digest)-1])
}

func mainWorkload() uint64 {
	var checksum uint64
	var crcBytes [4]byte
	for i := 0; i < 4000; i++ {
		data := payloads[i%len(payloads)]
		binary.BigEndian.PutUint32(crcBytes[:], crc32.ChecksumIEEE(data))
		checksum = foldDigest(checksum, crcBytes[:])

		md5Digest := md5.Sum(data)
		checksum = foldDigest(checksum, md5Digest[:])
		sha1Digest := sha1.Sum(data)
		checksum = foldDigest(checksum, sha1Digest[:])
		sha256Digest := sha256.Sum256(data)
		checksum = foldDigest(checksum, sha256Digest[:])

		mac := hmac.New(sha256.New, key)
		_, _ = mac.Write(data)
		checksum = foldDigest(checksum, mac.Sum(nil))
	}
	return checksum
}

func main() {
	fmt.Println(mainWorkload())
}
